controllers/upgrade.go

```diff
@@ -0,0 +1,43 @@
+package controllers
+
+import (
+	"context"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/replicatedhq/embedded-cluster-kinds/apis/v1beta1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// maybeSelfUpgrade checks if the embedded cluster version has changed and runs the upgrade command
+// if needed. If the version has changed, the process will exit.
+func (r *InstallationReconciler) maybeSelfUpgrade(ctx context.Context, in *v1beta1.Installation) {
+	curstr := strings.TrimPrefix(os.Getenv("EMBEDDEDCLUSTER_VERSION"), "v")
+	desstr := strings.TrimPrefix(in.Spec.Config.Version, "v")
```

Comment:
the old operator may not be able to parse the new installation CR. you’d want to only extract the version field without unmarshalling the entire CR.

Reply:
im not sure that is the case. k8s may handle [conversion](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/) and if it doesnt it probably wouldnt make it this far in the code